localstack/services/kms/models.py

```diff
@@ -0,0 +1,532 @@
+#   some state. It is modeled by KmsKey class.
+#
+# While KmsKeys always contain KmsCryptoKeys, sometimes KmsCryptoKeys exist without corresponding KmsKeys,
+# e.g. GenerateDataKeyPair API call returns contents of a new KmsCryptoKey that is not associated with any KmsKey,
+# but is partially encrypted by some pre-existing KmsKey.
+
+
+class KmsCryptoKey:
+    """
+    KmsCryptoKeys used to model both of the two cases where AWS generates keys:
+    1. Keys that are created to be used inside of AWS. For such a key, its key material / private key are not to
+    leave AWS unencrypted. If they have to leave AWS, a different KmsCryptoKey is used to encrypt the data first.
+    2. Keys that AWS creates for customers for some external use. Such a key might be returned to a customer with its
+    key material or public key unencrypted - see KMS GenerateDataKey / GenerateDataKeyPair. But such a key is not stored
+    by AWS and is not used by AWS.
+    """
+
+    def __init__(self, key_spec: str):
+        self.private_key = None
+        self.public_key = None
+        # Technically, key_material, being a symmetric encryption key, is only relevant for
+        #   key_spec == SYMMETRIC_DEFAULT.
+        # But LocalStack uses symmetric encryption with this key_material even for other specs. Asymmetric keys are
+        # generated, but are not actually used.
+        self.key_material = os.urandom(SYMMETRIC_DEFAULT_MATERIAL_LENGTH)
+
+        if key_spec == "SYMMETRIC_DEFAULT":
+            return
+
+        public_format = None
+        if key_spec.startswith("RSA"):
+            key_size = RSA_CRYPTO_KEY_LENGTHS.get(key_spec)
+            self.key = rsa.generate_private_key(public_exponent=65537, key_size=key_size)
+            public_format = crypto_serialization.PublicFormat.PKCS1
+        elif key_spec.startswith("ECC"):
+            curve = ECC_CURVES.get(key_spec)
+            self.key = ec.generate_private_key(curve)
+            public_format = crypto_serialization.PublicFormat.SubjectPublicKeyInfo
+        else:
+            # Currently we do not support HMAC keys - symmetric keys that are used for GenerateMac / VerifyMac.
+            # We also do not support SM2 - asymmetric keys both suitable for ENCRYPT_DECRYPT and SIGN_VERIFY,
+            # but only used in China AWS regions.
+            raise UnsupportedOperationException(f"KeySpec {key_spec} is not supported")
+
+        self.private_key = self.key.private_bytes(
+            crypto_serialization.Encoding.DER,
+            crypto_serialization.PrivateFormat.PKCS8,
+            crypto_serialization.NoEncryption(),
+        )
+        self.public_key = self.key.public_key().public_bytes(
+            crypto_serialization.Encoding.DER, public_format
+        )
+
+
+class KmsKey:
+    metadata: KeyMetadata
+    crypto_key: KmsCryptoKey
+
+    def __init__(
+        self, create_key_request: CreateKeyRequest, account_id: str = None, region: str = None
```

Comment:
nit: It seems a bit unusual to pass a "Create...Request" object to an object initializer - this seems like a good case for using a factory pattern with a static factory method, something like:
```
@staticmethod
def create_key(create_key_request: CreateKeyRequest, account_id: str = None, region: str = None) -> KmsKey
    key = ...
    return key
```
More food for thought - no change required for this initial version, but maybe we can tackle this in a follow-up iteration.

Reply:
Well, the request is just a bunch of settings for the initialization. It would still have to get there - if not as this dictionary, then in some other form. Adding one more method won't solve this.